aten/src/ATen/native/sparse/SparseFactories.cpp

```diff
@@ -0,0 +1,176 @@
+
+        for (const auto d_i : c10::irange(n_diag)) {
+          const int64_t& off_i = offsets_accessor[d_i];
+          const int64_t row_out_begin = off_i < 0 ? std::abs(off_i) : 0;
+          const int64_t col_out_begin = off_i > 0 ? off_i : 0;
+
+          // Number of column, row positions we can assign into and the number
+          // of diag positions we can read from
+          const int64_t row_slots = n_rows_out - row_out_begin;
+          const int64_t col_slots = n_cols_out - col_out_begin;
+          const int64_t diag_slots = n_cols_in - col_out_begin;
+
+          // Max number of reads lets us limit the loop by the first thing we
+          // will exhaust, note if any of these comes up negative irange will be
+          // empty so we don't have to handle that case explicitly
+          const int64_t max_read =
+              std::min(diag_slots, std::min(row_slots, col_slots));
+
+
+          for (const auto read_count : c10::irange(max_read)) {
+            const auto col = col_out_begin + read_count;
+            const auto row = row_out_begin + read_count;
+
+            values_accessor[actual_nnz] = diagonals_accessor[d_i][col];
+            indices_accessor[0][actual_nnz] = row;
+            indices_accessor[1][actual_nnz] = col;
+            ++actual_nnz;
+          }
+        }
+      });
+  indices = indices.narrow(1, 0, actual_nnz);
+  values = values.narrow(0, 0, actual_nnz);
+  SparseTensor sparse =
+      at::sparse_coo_tensor(indices, values, shape, sparse_options);
+  return sparse;
+}
+
+// Check offsets for duplicates, and out-of-bounds diagonals
+void validate_spdiags_offsets_cpu(
+    const Tensor& offsets,
+    const int64_t n_row,
+    const int64_t n_col) {
+  std::set<int64_t> seen_offsets;
+  auto offsets_accessor = offsets.accessor<int64_t, 1>();
+  for (auto i : c10::irange(offsets.size(0))) {
+    auto off = offsets_accessor[i];
+    TORCH_CHECK(
+        seen_offsets.insert(off).second,
+        "Offset array contains duplicate values");
+    TORCH_CHECK(
+        ((-1 * n_row) < off) && (off < n_col),
+        "Diagonal ",
+        off,
+        " does not exist in output shape (",
+        n_row,
+        ",",
+        n_col,
+        ")");
+  }
+}
+
+} // namespace
+
+SparseTensor spdiags_sparse_cpu(
+    const Tensor& diagonals,
+    const Tensor& offsets,
+    IntArrayRef shape,
+    c10::optional<Layout> layout) {
+  TORCH_CHECK(diagonals.dim() == 2, "Diagonals must be 2d");
+  TORCH_CHECK(shape.size() == 2, "Output shape must be 2d");
+  TORCH_CHECK(offsets.dim() == 1, "Offsets must be 1d");
+  TORCH_CHECK(
+      diagonals.size(0) == offsets.size(0),
+      "Number of diagonals (",
+      diagonals.size(0),
+      ") does not match the number of offsets (",
+      offsets.size(0),
+      ")");
+  if (layout) {
+    TORCH_CHECK(
+        (*layout == Layout::Sparse) || (*layout == Layout::SparseCsc) ||
+            (*layout == Layout::SparseCsr),
+        "Only output layouts (",
```

Comment:
@cpuhrsch , @IvanYashchuk , do we need to check that the input tensors are not sparse just to be extra safe?

Reply:
@nikitaved If the dispatcher enforces the correct device type should it not also enforce the layout? This function has no dispatch for `Sparse*` keys.

Reply:
Only for the first argument will it enforce afaik, not the others. index_select, for example, checks that the second argument representing index is dense.

Reply:
Ok, if that is the case then do I also need to check if the other tensor argument is also on the correct device? 

I am assuming by "check" you mean `TORCH_CHECK` that is, cause an error. I should not just silently convert the argument, right?

Reply:
Right, just an error message with `TORCH_CHECK` will do.